File: src/utils/track.ts
import { ArcSegment, BezierSegment, LineSegment, wrapAngle, XY } from "./math";

export type ClosestPointResult = {
    point: XY;
    tangent: XY;      // normalized track-forward direction at closest point
    distance: number;
    distanceSq: number;
    t: number;          // local segment parameter [0,1]
    distanceFromTrackPieceStart: number;
};

export type TrackSegment = LineSegment | ArcSegment | BezierSegment;

type ClosestPointOnTrackResult = ClosestPointResult & {
    segmentIndex: number;
};

export type TrackQueryResult = ClosestPointResult & {
    segmentIndex: number;
    lateralOffset: number; // right positive, left negative
    headingAngle: number;  // signed angle from track tangent to car heading, in [-PI, PI]
};

/** Creates a zero-initialized TrackQueryResult — use for pre-allocation, then pass to queryTrack. */
export function makeTrackQueryResult(): TrackQueryResult {
    return {
        point: { x: 0, y: 0 },
        tangent: { x: 0, y: 0 },
        distanceSq: 0, distance: 0, t: 0, distanceFromTrackPieceStart: 0,
        segmentIndex: -1, lateralOffset: 0, headingAngle: 0,
    };
}

// ---------------------------------------------------------------------------
// Module-level scratch objects — NEVER store these references; they are
// overwritten on every call. Safe because JS is single-threaded and the call
// chain is fully synchronous.
// ---------------------------------------------------------------------------
const _cpPt: XY = { x: 0, y: 0 };
const _cpTg: XY = { x: 0, y: 0 };
const _cpScratch: ClosestPointResult = {
    point: _cpPt, tangent: _cpTg,
    distanceSq: 0, distance: 0, t: 0, distanceFromTrackPieceStart: 0,
};
const _bestPt: XY = { x: 0, y: 0 };
const _bestTg: XY = { x: 0, y: 0 };
const _bestScratch: ClosestPointOnTrackResult = {
    point: _bestPt, tangent: _bestTg,
    distanceSq: Infinity, distance: 0, t: 0, distanceFromTrackPieceStart: 0,
    segmentIndex: 0,
};

/**
 * Signed angular travel from 'from' to 'to'.
 * If ccw=true, result is in [0, 2π).
 * If ccw=false, result is in (-2π, 0].
 */
export function signedArcDelta(from: number, to: number, ccw: boolean): number {
    let d = wrapAngle(to - from);

    if (ccw) {
        if (d < 0) d += 2 * Math.PI;
    } else {
        if (d > 0) d -= 2 * Math.PI;
    }

    return d;
}

// Returns _cpScratch — do NOT store the reference, read immediately.
export function closestPointOnLineSegment(seg: LineSegment, q: XY): ClosestPointResult {
    const ax = seg.start.x, ay = seg.start.y;
    const bx = seg.end.x,   by = seg.end.y;
    const abx = bx - ax,    aby = by - ay;
    const abLenSq = abx * abx + aby * aby;

    if (abLenSq < 1e-12) {
        const dx = q.x - ax, dy = q.y - ay;
        const d2 = dx * dx + dy * dy;
        _cpPt.x = ax; _cpPt.y = ay;
        _cpTg.x = 1;  _cpTg.y = 0;
        _cpScratch.distanceSq = d2;
        _cpScratch.distance = Math.sqrt(d2);
        _cpScratch.t = 0;
        _cpScratch.distanceFromTrackPieceStart = 0;
        return _cpScratch;
    }

    const raw = ((q.x - ax) * abx + (q.y - ay) * aby) / abLenSq;
    const t = raw < 0 ? 0 : raw > 1 ? 1 : raw;
    const px = ax + abx * t,  py = ay + aby * t;
    const abLen = Math.sqrt(abLenSq);
    const dx = q.x - px, dy = q.y - py;
    const d2 = dx * dx + dy * dy;

    _cpPt.x = px; _cpPt.y = py;
    _cpTg.x = abx / abLen; _cpTg.y = aby / abLen;
    _cpScratch.distanceSq = d2;
    _cpScratch.distance = Math.sqrt(d2);
    _cpScratch.t = t;
    _cpScratch.distanceFromTrackPieceStart = t * abLen;
    return _cpScratch;
}

// Returns _cpScratch — do NOT store the reference, read immediately.
export function closestPointOnArcSegment(seg: ArcSegment, q: XY): ClosestPointResult {
    const cx = seg.center.x, cy = seg.center.y;
    const vsx = seg.start.x - cx, vsy = seg.start.y - cy;
    const vex = seg.end.x   - cx, vey = seg.end.y   - cy;
    const vqx = q.x - cx,         vqy = q.y - cy;

    const rs = Math.sqrt(vsx * vsx + vsy * vsy);
    const re = Math.sqrt(vex * vex + vey * vey);

    if (rs < 1e-12 || re < 1e-12) throw new Error("Arc start/end must not coincide with center.");

    const radius = 0.5 * (rs + re);
    if (Math.abs(rs - re) > 1e-6 * Math.max(1, radius)) throw new Error("Arc start and end are not on the same circle.");

    const a0 = Math.atan2(vsy, vsx);
    const a1 = Math.atan2(vey, vex);
    const aq = Math.atan2(vqy, vqx);

    const useCCW = seg.clockwise;
    const totalDelta = signedArcDelta(a0, a1, useCCW);
    const qDelta = signedArcDelta(a0, aq, useCCW);
    const onArc = useCCW
        ? qDelta >= 0 && qDelta <= totalDelta
        : qDelta <= 0 && qDelta >= totalDelta;

    let px: number, py: number, t: number;
    const vqLenSq = vqx * vqx + vqy * vqy;

    if (vqLenSq === 0) {
        const dsx = q.x - seg.start.x, dsy = q.y - seg.start.y;
        const dex = q.x - seg.end.x,   dey = q.y - seg.end.y;
        if (dsx * dsx + dsy * dsy <= dex * dex + dey * dey) {
            px = seg.start.x; py = seg.start.y; t = 0;
        } else {
            px = seg.end.x;   py = seg.end.y;   t = 1;
        }
    } else if (onArc) {
        const vqLen = Math.sqrt(vqLenSq);
        px = cx + (vqx / vqLen) * radius;
        py = cy + (vqy / vqLen) * radius;
        t = totalDelta === 0 ? 0 : qDelta / totalDelta;
    } else {
        const dsx = q.x - seg.start.x, dsy = q.y - seg.start.y;
        const dex = q.x - seg.end.x,   dey = q.y - seg.end.y;
        if (dsx * dsx + dsy * dsy <= dex * dex + dey * dey) {
            px = seg.start.x; py = seg.start.y; t = 0;
        } else {
            px = seg.end.x;   py = seg.end.y;   t = 1;
        }
    }

    // Forward tangent: rotate unit radial by ±90° (already unit length).
    const rpx = px - cx, rpy = py - cy;
    const rpLen = Math.sqrt(rpx * rpx + rpy * rpy);
    let tanX: number, tanY: number;
    if (rpLen < 1e-12) { tanX = 1; tanY = 0; }
    else {
        const rx = rpx / rpLen, ry = rpy / rpLen;
        if (!seg.clockwise) { tanX = ry;  tanY = -rx; }  // rotate -90°
        else                { tanX = -ry; tanY =  rx; }  // rotate +90°
    }

    const dx = q.x - px, dy = q.y - py;
    const d2 = dx * dx + dy * dy;
    const tc = t < 0 ? 0 : t > 1 ? 1 : t;

    _cpPt.x = px; _cpPt.y = py;
    _cpTg.x = tanX; _cpTg.y = tanY;
    _cpScratch.distanceSq = d2;
    _cpScratch.distance = Math.sqrt(d2);
    _cpScratch.t = tc;
    _cpScratch.distanceFromTrackPieceStart = tc * Math.abs(totalDelta) * radius;
    return _cpScratch;
}

/**
 * Right normal for a forward tangent.
 * If tangent points forward, this points to track-right.
 */
function rightNormalFromTangent(tangent: XY): XY {
    return { x: tangent.y, y: -tangent.x };
}

function carDirectionFromAngle(angle: number): XY {
    return { x: Math.cos(angle), y: Math.sin(angle) };
}

function closestPointOnSegment(seg: TrackSegment, q: XY): ClosestPointResult {
    switch (seg.kind) {
        case "line":
            return closestPointOnLineSegment(seg, q);
        case "arc":
            return closestPointOnArcSegment(seg, q);
        default:
            throw new Error("Unsupported segment kind: " + (seg as any).kind);
        // case "bezier":
        //   return closestPointOnBezier(seg, q);
    }
}

// Search radius for neighbour-hint queries: checks hint ± NEIGHBOR_SEARCH_RADIUS segments.
// A car travelling at 80 m/s × 1/60 s × UNITS_PER_METER fits well within 1 segment, so
// radius 2 gives a comfortable safety margin while keeping the scan to 5 segments max.
const NEIGHBOR_SEARCH_RADIUS = 2;

// Returns _bestScratch — do NOT store the reference, read immediately.
function findClosestOnTrack(
    track: TrackSegment[],
    carPos: XY,
    hint: number
): ClosestPointOnTrackResult {
    if (hint >= 0 && hint < track.length && track.length > NEIGHBOR_SEARCH_RADIUS * 2) {
        let found = false;
        for (let offset = -NEIGHBOR_SEARCH_RADIUS; offset <= NEIGHBOR_SEARCH_RADIUS; offset++) {
            const i = (hint + offset + track.length) % track.length;
            const r = closestPointOnSegment(track[i], carPos); // writes _cpScratch
            if (!found || r.distanceSq < _bestScratch.distanceSq) {
                _bestPt.x = r.point.x;   _bestPt.y = r.point.y;
                _bestTg.x = r.tangent.x; _bestTg.y = r.tangent.y;
                _bestScratch.distanceSq = r.distanceSq;
                _bestScratch.distance   = r.distance;
                _bestScratch.t          = r.t;
                _bestScratch.distanceFromTrackPieceStart = r.distanceFromTrackPieceStart;
                _bestScratch.segmentIndex = i;
                found = true;
            }
        }
        if (found && _bestScratch.distanceSq < 1e9) return _bestScratch;
    }

    // Full scan fallback.
    let found = false;
    for (let i = 0; i < track.length; i++) {
        const r = closestPointOnSegment(track[i], carPos); // writes _cpScratch
        if (!found || r.distanceSq < _bestScratch.distanceSq) {
            _bestPt.x = r.point.x;   _bestPt.y = r.point.y;
            _bestTg.x = r.tangent.x; _bestTg.y = r.tangent.y;
            _bestScratch.distanceSq = r.distanceSq;
            _bestScratch.distance   = r.distance;
            _bestScratch.t          = r.t;
            _bestScratch.distanceFromTrackPieceStart = r.distanceFromTrackPieceStart;
            _bestScratch.segmentIndex = i;
            found = true;
        }
    }
    return _bestScratch;
}

/**
 * Query the track for the car's position. Writes the result into `out` if provided
 * (recommended — pass a pre-allocated TrackQueryResult to avoid heap allocation).
 * If `out` is omitted a new object is allocated (first-call / non-hot-path use only).
 */
export function queryTrack(
    track: TrackSegment[],
    carPos: XY,
    carAngle: number,
    hintSegmentIndex = -1,
    out?: TrackQueryResult
): TrackQueryResult {
    if (track.length === 0) throw new Error("Track must contain at least one segment.");

    const best = findClosestOnTrack(track, carPos, hintSegmentIndex); // writes _bestScratch

    // Normalize tangent (inline — avoids sub/normalize allocations)
    const tx = best.tangent.x, ty = best.tangent.y;
    const tLen = Math.sqrt(tx * tx + ty * ty);
    const ntx = tLen > 1e-12 ? tx / tLen : tx;
    const nty = tLen > 1e-12 ? ty / tLen : ty;

    // delta = carPos - best.point
    const dx = carPos.x - best.point.x;
    const dy = carPos.y - best.point.y;

    // right = rightNormal(tangent) = { x: nty, y: -ntx }
    const lateralOffset = dx * nty + dy * (-ntx);

    const headingAngle = wrapAngle(carAngle - Math.atan2(nty, ntx));

    const result = out ?? makeTrackQueryResult();
    result.point.x   = best.point.x;
    result.point.y   = best.point.y;
    result.tangent.x = ntx;
    result.tangent.y = nty;
    result.distanceSq = best.distanceSq;
    result.distance   = best.distance;
    result.t          = best.t;
    result.distanceFromTrackPieceStart = best.distanceFromTrackPieceStart;
    result.segmentIndex  = best.segmentIndex;
    result.lateralOffset = lateralOffset;
    result.headingAngle  = headingAngle;
    return result;
}

File: csharp/Utils/Track.cs
using System;

namespace SmartRace.Utils
{
    public struct XY
    {
        public double X { get; set; }
        public double Y { get; set; }

        public XY(double x, double y)
        {
            X = x;
            Y = y;
        }

        public override string ToString() => $"({X}, {Y})";
    }

    public struct ClosestPointResult
    {
        public XY Point { get; set; }
        public XY Tangent { get; set; }       // normalized track-forward direction at closest point
        public double Distance { get; set; }
        public double DistanceSq { get; set; }
        public double T { get; set; }         // local segment parameter [0,1]
        public double DistanceFromTrackPieceStart { get; set; }
    }

    public struct LineSegment
    {
        public string Kind { get; }
        public XY Start { get; set; }
        public XY End { get; set; }

        public LineSegment(XY start, XY end)
        {
            Kind = "line";
            Start = start;
            End = end;
        }
    }

    public struct ArcSegment
    {
        public string Kind { get; }
        public XY Start { get; set; }
        public XY End { get; set; }
        public XY Center { get; set; }
        public bool Clockwise { get; set; }

        public ArcSegment(XY start, XY end, XY center, bool clockwise)
        {
            Kind = "arc";
            Start = start;
            End = end;
            Center = center;
            Clockwise = clockwise;
        }
    }

    public struct BezierSegment
    {
        public string Kind { get; }
        public XY P0 { get; set; }
        public XY P1 { get; set; }
        public XY P2 { get; set; }
        public XY P3 { get; set; }

        public BezierSegment(XY p0, XY p1, XY p2, XY p3)
        {
            Kind = "bezier";
            P0 = p0;
            P1 = p1;
            P2 = p2;
            P3 = p3;
        }
    }

    public interface ITrackSegment
    {
        string Kind { get; }
    }

    public struct TrackSegment : ITrackSegment
    {
        private readonly object _segment;
        
        public string Kind { get; private set; }

        public TrackSegment(LineSegment segment)
        {
            _segment = segment;
            Kind = "line";
        }

        public TrackSegment(ArcSegment segment)
        {
            _segment = segment;
            Kind = "arc";
        }

        public TrackSegment(BezierSegment segment)
        {
            _segment = segment;
            Kind = "bezier";
        }

        public LineSegment AsLine() => (LineSegment)_segment;
        public ArcSegment AsArc() => (ArcSegment)_segment;
        public BezierSegment AsBezier() => (BezierSegment)_segment;
    }

    public struct ClosestPointOnTrackResult
    {
        public XY Point { get; set; }
        public XY Tangent { get; set; }
        public double Distance { get; set; }
        public double DistanceSq { get; set; }
        public double T { get; set; }
        public int SegmentIndex { get; set; }
        public double DistanceFromTrackPieceStart { get; set; }
    }

    public struct TrackQueryResult
    {
        public XY Point { get; set; }
        public XY Tangent { get; set; }
        public double Distance { get; set; }
        public double DistanceSq { get; set; }
        public double T { get; set; }
        public int SegmentIndex { get; set; }
        public double LateralOffset { get; set; }    // right positive, left negative
        public double HeadingAngle { get; set; }     // signed angle from track tangent to car heading, in [-PI, PI]
        public double DistanceFromTrackPieceStart { get; set; }
    }

    public static class TrackMath
    {
        public static XY Add(XY a, XY b)
        {
            return new XY(a.X + b.X, a.Y + b.Y);
        }

        public static XY Sub(XY a, XY b)
        {
            return new XY(a.X - b.X, a.Y - b.Y);
        }

        public static XY Mul(XY v, double s)
        {
            return new XY(v.X * s, v.Y * s);
        }

        public static double Dot(XY a, XY b)
        {
            return a.X * b.X + a.Y * b.Y;
        }

        public static double Cross(XY a, XY b)
        {
            return a.X * b.Y - a.Y * b.X;
        }

        public static double LengthSq(XY v)
        {
            return Dot(v, v);
        }

        public static double Length(XY v)
        {
            return Math.Sqrt(LengthSq(v));
        }

        public static XY Normalize(XY v)
        {
            double len = Length(v);
            if (len == 0) return new XY(0, 0);
            return new XY(v.X / len, v.Y / len);
        }

        public static double Clamp(double v, double min, double max)
        {
            return Math.Max(min, Math.Min(max, v));
        }

        public static double DistanceSq(XY a, XY b)
        {
            return LengthSq(Sub(a, b));
        }

        public static double AngleOf(XY v)
        {
            return Math.Atan2(v.Y, v.X);
        }

        public static double WrapAngle(double a)
        {
            while (a <= -Math.PI) a += 2 * Math.PI;
            while (a > Math.PI) a -= 2 * Math.PI;
            return a;
        }

        /// <summary>
        /// Signed angular travel from 'from' to 'to'.
        /// If ccw=true, result is in [0, 2π).
        /// If ccw=false, result is in (-2π, 0].
        /// </summary>
        public static double SignedArcDelta(double from, double to, bool ccw)
        {
            double d = WrapAngle(to - from);

            if (ccw)
            {
                if (d < 0) d += 2 * Math.PI;
            }
            else
            {
                if (d > 0) d -= 2 * Math.PI;
            }

            return d;
        }

        public static ClosestPointResult ClosestPointOnLineSegment(LineSegment seg, XY q)
        {
            XY a = seg.Start;
            XY b = seg.End;
            XY ab = Sub(b, a);
            double abLenSq = LengthSq(ab);

            if (abLenSq < 1e-12)
            {
                return new ClosestPointResult
                {
                    Point = a,
                    Tangent = new XY(1, 0),
                    DistanceSq = (double)DistanceSq(q, a),
                    Distance = Math.Sqrt((double)DistanceSq(q, a)),
                    T = 0,
                    DistanceFromTrackPieceStart = 0,
                };
            }

            double t = Clamp(Dot(Sub(q, a), ab) / abLenSq, 0, 1);
            XY point = Add(a, Mul(ab, t));
            XY tangent = Normalize(ab);
            double d2 = DistanceSq(q, point);

            return new ClosestPointResult
            {
                Point = point,
                Tangent = tangent,
                DistanceSq = d2,
                Distance = Math.Sqrt(d2),
                T = t,
                DistanceFromTrackPieceStart = t * Math.Sqrt(abLenSq),
            };
        }

        public static ClosestPointResult ClosestPointOnArcSegment(ArcSegment seg, XY q)
        {
            XY vs = Sub(seg.Start, seg.Center);
            XY ve = Sub(seg.End, seg.Center);
            XY vq = Sub(q, seg.Center);

            double rs = Length(vs);
            double re = Length(ve);

            if (rs < 1e-12 || re < 1e-12)
            {
                throw new InvalidOperationException("Arc start/end must not coincide with center.");
            }

            double radius = 0.5 * (rs + re);
            if (Math.Abs(rs - re) > 1e-6 * Math.Max(1, radius))
            {
                throw new InvalidOperationException("Arc start and end are not on the same circle.");
            }

            double a0 = AngleOf(vs);
            double a1 = AngleOf(ve);
            double aq = AngleOf(vq);

            bool useCCW = seg.Clockwise;
            double totalDelta = SignedArcDelta(a0, a1, useCCW);

            // Where does q lie along that directed arc?
            double qDelta = SignedArcDelta(a0, aq, useCCW);
            bool onArc = useCCW
                ? qDelta >= 0 && qDelta <= totalDelta
                : qDelta <= 0 && qDelta >= totalDelta;

            XY point;
            double t;

            if (LengthSq(vq) == 0)
            {
                // Query exactly at center: any circle point is equally radially valid,
                // so fall back to nearest endpoint.
                double d2Start = DistanceSq(q, seg.Start);
                double d2End = DistanceSq(q, seg.End);
                if (d2Start <= d2End)
                {
                    point = seg.Start;
                    t = 0;
                }
                else
                {
                    point = seg.End;
                    t = 1;
                }
            }
            else if (onArc)
            {
                XY dir = Normalize(vq);
                point = Add(seg.Center, Mul(dir, radius));
                t = totalDelta == 0 ? 0 : qDelta / totalDelta;
            }
            else
            {
                double d2Start = DistanceSq(q, seg.Start);
                double d2End = DistanceSq(q, seg.End);
                if (d2Start <= d2End)
                {
                    point = seg.Start;
                    t = 0;
                }
                else
                {
                    point = seg.End;
                    t = 1;
                }
            }

            // Forward tangent follows arc direction.
            XY radial = Normalize(Sub(point, seg.Center));
            XY tangent = !seg.Clockwise
                ? new XY(radial.Y, -radial.X)   // rotate radial by -90°
                : new XY(-radial.Y, radial.X);  // rotate radial by +90°

            double d2 = DistanceSq(q, point);

            double distanceFromTrackPieceStart = t * Math.Abs(totalDelta) * radius;

            return new ClosestPointResult
            {
                Point = point,
                Tangent = Normalize(tangent),
                DistanceSq = d2,
                Distance = Math.Sqrt(d2),
                T = Clamp(t, 0, 1),
                DistanceFromTrackPieceStart = distanceFromTrackPieceStart,
            };
        }

        /// <summary>
        /// Right normal for a forward tangent.
        /// If tangent points forward, this points to track-right.
        /// </summary>
        private static XY RightNormalFromTangent(XY tangent)
        {
            return new XY(tangent.Y, -tangent.X);
        }

        private static XY CarDirectionFromAngle(double angle)
        {
            return new XY(Math.Cos(angle), Math.Sin(angle));
        }

        public static ClosestPointResult ClosestPointOnSegment(TrackSegment seg, XY q)
        {
            switch (seg.Kind)
            {
                case "line":
                    return ClosestPointOnLineSegment(seg.AsLine(), q);
                case "arc":
                    return ClosestPointOnArcSegment(seg.AsArc(), q);
                default:
                    throw new NotSupportedException($"Unsupported segment kind: {seg.Kind}");
                // case "bezier":
                //   return ClosestPointOnBezier(seg, q);
            }
        }

        public static TrackQueryResult QueryTrack(TrackSegment[] track, XY carPos, double carAngle)
        {
            if (track.Length == 0)
            {
                throw new ArgumentException("Track must contain at least one segment.");
            }

            ClosestPointOnTrackResult? best = null;

            for (int i = 0; i < track.Length; i++)
            {
                ClosestPointResult r = ClosestPointOnSegment(track[i], carPos);
                if (!best.HasValue || r.DistanceSq < best.Value.DistanceSq)
                {
                    best = new ClosestPointOnTrackResult
                    {
                        Point = r.Point,
                        Tangent = r.Tangent,
                        Distance = r.Distance,
                        DistanceSq = r.DistanceSq,
                        T = r.T,
                        SegmentIndex = i,
                        DistanceFromTrackPieceStart = r.DistanceFromTrackPieceStart,
                    };
                }
            }

            XY tangent = Normalize(best.Value.Tangent);
            XY delta = Sub(carPos, best.Value.Point);

            // Right-positive lateral offset
            XY right = RightNormalFromTangent(tangent);
            double lateralOffset = Dot(delta, right);

            // Signed heading angle from track tangent to car heading, in [-PI, PI]
            double trackAngle = AngleOf(tangent);
            double headingAngle = WrapAngle(carAngle - trackAngle);

            return new TrackQueryResult
            {
                Point = best.Value.Point,
                Tangent = tangent,
                Distance = best.Value.Distance,
                DistanceSq = best.Value.DistanceSq,
                T = best.Value.T,
                SegmentIndex = best.Value.SegmentIndex,
                LateralOffset = lateralOffset,
                HeadingAngle = headingAngle,
                DistanceFromTrackPieceStart = best.Value.DistanceFromTrackPieceStart,
            };
        }
    }
}